### PR TITLE
APP-456: Query multi action variable, fetch and cast stored multi action values

### PR DESCRIPTION
**Pull request by hugokallstrom**
This got quite messy, suggestion on improvements very welcome.

### Checklist

- [x] Functionality is covered by a unit test
- [x] Functionality is accessible in engineering mode

### CODE REVIEW

File: app/build.gradle.kts [hunk 1]
```diff
@@ -149,6 +149,7 @@ dependencies {
     androidTestImplementation(Libs.AndroidX.Test.runner)
     androidTestImplementation(Libs.AndroidX.Test.rules)
     androidTestImplementation(Libs.AndroidX.Test.junit)
+    testImplementation("org.json:json:20210307")
```
Comment (hugokallstrom) [inline app/build.gradle.kts:153]: Had to include this in order to use JSONObject in unit tests

File: app/src/main/java/com/hedvig/app/feature/embark/util/VariableExtractor.kt [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            extractMultiActionVariables(it, valueStore)
+        }
+
+        multiActionVariables.forEach {
+            regularVariables = regularVariables.plus(it)
+        }
+        return regularVariables
+    }
+
+    private fun extractRegularVariables(
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/embark/util/VariableExtractor.kt:29]: Could we split this into two methods, `extractSingleVariable`/`extractGeneratedVariable` which as input take `GraphQLVariablesFragment.AsGraphQLSingleVariable`/`GraphQLVariablesFragment.AsGeneratedVariable`? That would remove the need for having a possible null return for this/these functions.

File: app/src/main/java/com/hedvig/app/feature/embark/util/VariableExtractor.kt [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        return null
+    }
+
+    private fun extractMultiActionVariables(
+        variable: GraphQLVariablesFragment,
+        valueStore: ValueStore
+    ): JSONObject? {
+        variable.asEmbarkAPIGraphQLMultiActionVariable?.let { multiActionVariables ->
+            val multiActionItems = valueStore.getMultiActionItems(multiActionVariables.key)
+            val casted = multiActionItems.map { multiActionMap ->
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/embark/util/VariableExtractor.kt:57]: If we squint a little at this function, it looks like it does the same job as `extractVariables` does at top-level, but with a `ValueStore` scoped to the Multi Action-item. Perhaps we could make that the reality in code, as well? Then we would not have to express the procedure twice, so to speak.

We could accomplish this by translating the `GraphQLVariablesFragment.AsGraphQLMultiActionVariable` into a `GraphQLVariablesFragment`, creating a static `ValueStore` from the `multiActionItems`, and then finally calling `extractVariables` with those two parameters as the argument.

What are your thoughts on that approach?

File: app/src/test/java/com/hedvig/app/feature/embark/VariableExtractorTest.kt [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+import assertk.assertThat
+import assertk.assertions.isEqualTo
+import com.hedvig.android.owldroid.fragment.GraphQLVariablesFragment
+import com.hedvig.android.owldroid.type.EmbarkAPIGraphQLSingleVariableCasting
+import com.hedvig.app.feature.embark.util.VariableExtractor
+import org.json.JSONObject
+import org.junit.Test
+
+class VariableExtractorTest {
```
Comment (hedvigoscar) [inline app/src/test/java/com/hedvig/app/feature/embark/VariableExtractorTest.kt:11]: 🚀

**Review by hedvigoscar — COMMENTED** [review]
Looks mostly good I think!

File: app/src/main/java/com/hedvig/app/feature/embark/util/VariableExtractor.kt (hunk 3 above)
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/feature/embark/util/VariableExtractor.kt:57]: Yes I was thinking the same, but had difficulty creating the fragment in `embarkStory.graphql` - how would that look? Thanks

**Review by hedvigoscar — APPROVED** [review]

**Comment by hedvigoscar** [comment]
Please address the failing test before merging, though 🤔

**Pull request merged**